### ISSUE TITLE
Reset Ring after background thread task is stopped / finished

### ISSUE THREAD

**Issue by dan-burt** [luispadron/UICircularProgressRing#72]
- Version = 1.7.5

I am using the Ring in a dynamic fashion that constantly changes based on CoreML processing from the camera (rather than a start-to-finish task resulting in the ring reaching 100%). The user initiates and then stops the processing. The processing occurs on a background thread (I believe I can specify which QoS but have had no need to use any other than the default).

When the user stops the processing, I need to reset the UI. I would like the Ring to reset itself back to 0.0, and a slower animation would be great. So in the section of the button tap, I call `setProgress(value: 0.0, animationDuration: 0.5)` function. I have also encapsulated this call within a `DispatchQueue.main.async { }` call as well, but it behaves the same.

However, most of the time, it remains stuck at some other number.

I have seen there is a completion handler, but assumed this was when a task "naturally" completed, rather than a button tap by the user. Or are you able to show any simple demonstrations of the completion handler, if I have misinterpreted this?

**Comment by luispadron** [comment]
>  I have also encapsulated this call within a DispatchQueue.main.async { } call as well, but it behaves the same

What exactly is the behavior you're experiencing? 

If you could post sample code of what you're doing that would be easier to debug, thanks.

**Comment by dan-burt** [comment]
the user starts the camera, and most processing is within the `captureOutput` delegate function (called after each frame is taken by the camera):

```
func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
     // some other code happens here
     DispatchQueue.main.async {
                self.ringProgress.setProgress(value: CGFloat(result), animationDuration: 0.2)
     }
     // other code
}
```

when the user stops the camera, i have a simple tidy up function:

func cameraRelease() {
    cameraSession.stopRunning()
    ringProgress.setProgress(value: CGFloat(0.0), animationDuration: 0.5)
}

This last call has been tried within the DispatchQueue.main.async { } call.

Hope it explains better...

**Comment by luispadron** [comment]
And you're saying that 

```swift
func cameraRelease() {
    cameraSession.stopRunning()
    ringProgress.setProgress(value: CGFloat(0.0), animationDuration: 0.5)
}
```

Doesn't cause the ring to be reset?

**Comment by dan-burt** [comment]
No, its like its finishing up the queue of frames from the background task, or just leaves it at the last value that it was set to (not sure which, nor how you could find out).

**Comment by dan-burt** [comment]
More code.... This process is initiated under:

```
cameraOutput.setSampleBufferDelegate(self, queue: DispatchQueue(label: "videoQueue"))
cameraSession.startRunning()
```

Can I access this queue? So I can force it to end, or supply a closure or other functionality?

**Comment by luispadron** [comment]
I’m pretty confused currently, you’re saying the ring isn’t reset?

The documentation states that calling `setProgress` while ring is already animating will stop the current animation and start a new one. If you call `setProgress` with value 0, depending on what your min value is, will reset the ring back to 0. This needs to happen on the main thread which can be done by using 

```swift
DispatchQueue.main.async {
    // code here
}
```

I recommend reading the documentation if you need more information. If you’re trying to report a bug could you post a video or an example project reproducing the bug.

**Comment by dan-burt** [comment]
The user starts the camera. This kicks off some functions to initialise the camera using AVFoundation, starts an AVCaptureSession, then using the captureOutput(didOutput: from:) method it runs some code for each frame captured (I assume many, 25fps?, per second).

For each frame, it should update the ring. This works as expected. Though as you say, the next setProgress is probably overriding the previous calls on a number of occasions.

The user then stops the camera. This should reset the AVCaptureSession and other stuff including the UI. This is already executed on the main thread. I mentioned this in my posts only to demonstrate I have tried this. My `cameraRelease()` function is a standard function contained within my ViewController class, no queues involved.

My last question(s) was purely a "what if" and I know is not within the bounds of your UICircularProgressRing functionality.

I have also tried putting in a sleep() call before the final `setProgress()` call but this still does not do anything... This is what made me think of queues.

My apologies, but I am a newcomer to iOS programming, so not sure of advanced debugging techniques, nor intricacies of the Swift language...

**Comment by luispadron** [comment]
Not a problem, we all start somewhere. 

Since you mentioned the ring doesn’t go fully to 0, but stops at some other number entirely, this leads me to believe that after you reset the other queue isn’t done being processed. I would recommend putting breakpoints whenever you call setProgress and verify the call times are as expected, or even easier put some print statements right before and see if anything weird is happening there. Queues are tricky so I understand the pain. 

If you can reproduce the behavior or send me an actual sample I can debug it further myself as well.

**Comment by dan-burt** [comment]
thanks for the advice, Luis. Using breakpoints, and some other debugging, I found that after the `stopRunning()` call, there was indeed still some values set and the processCaptureFrame function running (on another / background thread).

to resolve, and maybe not the most elegant, I had to add another Bool var to toggle between the `stopRunning()` and `startRunning()` calls. At the end of `captureOutput()` function, I test this Bool and force the `setProgress()` here if it is `false` as I known it should not initiate another execution of this function.

the UICircularProgressRing was functioning as expected, apologies for creating an "issue" on your timeline when in the end there wasn't one.
